pkg/reconciler/internal/paramvalidation/paramvalidation.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2023 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package paramvalidation
+
+import (
+	"fmt"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"github.com/tektoncd/pipeline/pkg/list"
+	"github.com/tektoncd/pipeline/pkg/substitution"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// MissingKeysObjectParamNames checks if all required keys of object type params are provided in taskrun params or taskSpec's default.
+func MissingKeysObjectParamNames(paramSpecs []v1beta1.ParamSpec, params []v1beta1.Param) map[string][]string {
+	neededKeys := make(map[string][]string)
+	providedKeys := make(map[string][]string)
+
+	for _, spec := range paramSpecs {
+		if spec.Type == v1beta1.ParamTypeObject {
+			// collect required keys from properties section
+			for key := range spec.Properties {
+				neededKeys[spec.Name] = append(neededKeys[spec.Name], key)
+			}
+
+			// collect provided keys from default
+			if spec.Default != nil && spec.Default.ObjectVal != nil {
+				for key := range spec.Default.ObjectVal {
+					providedKeys[spec.Name] = append(providedKeys[spec.Name], key)
+				}
+			}
+		}
+	}
+
+	// collect provided keys from run level value
+	for _, p := range params {
+		if p.Value.Type == v1beta1.ParamTypeObject {
+			for key := range p.Value.ObjectVal {
+				providedKeys[p.Name] = append(providedKeys[p.Name], key)
+			}
+		}
+	}
+
+	return findMissingKeys(neededKeys, providedKeys)
+}
+
+// findMissingKeys checks if objects have missing keys in its providers (taskrun value and default)
+func findMissingKeys(neededKeys, providedKeys map[string][]string) map[string][]string {
+	missings := map[string][]string{}
+	for p, keys := range providedKeys {
+		if _, ok := neededKeys[p]; !ok {
+			// Ignore any missing objects - this happens when object param is provided with default
+			continue
+		}
+		missedKeys := list.DiffLeft(neededKeys[p], keys)
+		if len(missedKeys) != 0 {
+			missings[p] = missedKeys
+		}
+	}
+
+	return missings
+}
+
+// ExtractParamArrayLengths extract and return the lengths of all array params
+// Example of returned value: {"$(params.array-params)": 2, "$(params[\"array-params\"])": 2, "$(params['array-params'])": 2}
```

Comment:
I'm a bit confused about this example, aren't these all the same param but with different syntax referenced? would it make sense to return a map of param name -> length? Consider also noting in the docstring that the only keys present in the output map are for array params, regardless of what was passed in, and that this doesn't return an error if a param doesn't have a default or provided value.

Reply:
> I'm a bit confused about this example, aren't these all the same param but with different syntax referenced? would it make sense to return a map of param name -> length?

Yes I also thought about that way. I think that is a cleaner solution. But I didn't change the logic b/c I want to keep the PR with as less changes as possible. It could be an improvement. Maybe I can open an issue to improve it?

Reply:
> Consider also noting in the docstring that the only keys present in the output map are for array params, regardless of what was passed in, and that this doesn't return an error if a param doesn't have a default or provided value.

Yes, we only check the array params, if for the array param both are missing then the output map will be empty, will also return error later

Reply:
I think since this is a refactoring PR where you're rewriting this function, I would make any readability updates you want to make to the function in this PR. From a quick skim of this PR, it looks like the functions that would need updates are all in this file, so I'm not sure that it would increase the PR size very much, but please correct me if I'm wrong.

Re: the docstring, can you please add that info there?

Reply:
Ok, let me try to rewrite this part.

Reply:
I think we can rewrite it, found out one function in substitution which can help to extract the param name from different formats of refs.
I have updated the functions.

Reply:
thanks! this looks better!